053-configurable-event-store/proposal.md

```diff
@@ -0,0 +1,457 @@
+  `build.SaveEvent(event)` is called in order to write build events to the
+  external store. The reason for accepting a list of `events` (despite
+  `db.Build.SaveEvent` only ever passing in a single event) is because if we
+  want to migrate build events to another `EventStore`, it would be super slow
+  making make a huge number of small `Put` calls when we can do one batch `Put`
+  per Build.
+
+* `Get(build db.Build, requested int, cursor *Key)` fetches events from the
+  `EventStore`, starting from an initial `Key` (which is excluded from the
+  result set - e.g.  if `cursor` points to the 10th `Key`, the 11th event will be
+  the next one fetched). `Key` can be any type, and is implementation-specific.
+  For `Postgres`, this will likely be something like `uint`, but other stores
+  (e.g. DynamoDB) that don't provide easy support for auto-incrementing IDs may
+  choose a different type of `Key`. Regardless of the type of `Key`, if the goal
+  is to fetch all events, `cursor` should initially point to `nil` (but the
+  pointer itself must not be `nil`). For instance, you can iterate over all
+  events for a completed build like so:
+
+  ```go
+  var cursor Key
+  batchSize := 1000
+  for {
+      events, err := eventStore.Get(build, batchSize, &cursor)
+      if err != nil {
+          panic(err)
+      }
+      // process events here
+      if len(events) < batchSize {
+          // there are no more events (see note below about `requested`)
+          break
+      }
+  }
+  ```
+
+  However, when possible, consumers of these build events should prefer to use
+  `db.Build.Events()`, which queries the `EventStore` and subscribes to a
+  notification bus to allow tracking in-progress builds (more on this in
+  [Notifications](#notifications)).
+
+  `Get` also takes in `requested` number of events to fetch. This is a soft
+  limit. To explain what I mean by this, suppose there are `n` events available
+  in the `EventStore` (after the `cursor`):
+  * If `n < requested`, `Get(...)` should return all `n` events
+  * If `n >= requested`, `Get(...)` should return at least `requested` events,
+    but may return more than `requested` if convenient (i.e. it may return
+    `requested <= x <= n`). If the backend store doesn't provide an easy way to
+    read exactly `requested` elements, and the `EventStore` reads in a chunk of
+    data that results in more than `requested` events, there's no use in throwing
+    the extra events away.
+
+* `Delete(builds []db.Builds)` deletes the build events for a list of builds.
+  This is used by the build log collector.
+
+* `DeletePipeline(pipeline db.Pipeline)` deletes all of the builds for a given
+  pipeline. While this could be done by doing something like:
+
+  ```go
+  builds, _ := pipeline.Builds()
+  eventStore.Delete(builds)
+  ```
+
+  ...this doesn't work so well for the Postgres implementation. Currently, when
+  we delete a Pipeline from the database, `TRIGGER`s will `DROP` the
```

Comment:
we could still `drop` even though not a from a trigger, right? I mean, from a regular `Exec`.

(you probably noticed, but, just for transparency: these are the triggers we currently have:

```
 table_schema | table_name | trigger_schema |             trigger_name             | event  | activation | condition |              definition
--------------+------------+----------------+--------------------------------------+--------+------------+-----------+---------------------------------------
 public       | pipelines  | public         | pipeline_build_events_delete_trigger | DELETE | AFTER      |           | EXECUTE FUNCTION on_pipeline_delete()
 public       | pipelines  | public         | pipeline_build_events_insert_trigger | INSERT | AFTER      |           | EXECUTE FUNCTION on_pipeline_insert()
 public       | teams      | public         | team_build_events_delete_trigger     | DELETE | AFTER      |           | EXECUTE FUNCTION on_team_delete()
 public       | teams      | public         | team_build_events_insert_trigger     | INSERT | AFTER      |           | EXECUTE FUNCTION on_team_insert()
```
)

where

```
on_team_delete
    EXECUTE format('DROP TABLE IF EXISTS team_build_events_%s', OLD.id);RETURN NULL;


on_team_insert
  EXECUTE format('CREATE TABLE IF NOT EXISTS team_build_events_%s () INHERITS (build_events)', NEW.id);RETURN NULL;


on_pipeline_delete
  EXECUTE format('DROP TABLE IF EXISTS pipeline_build_events_%s', OLD.id);RETURN NULL;


on_pipeline_insert
  EXECUTE format('CREATE TABLE IF NOT EXISTS pipeline_build_events_%s () 
        INHERITS (build_events)', NEW.id);
  
  EXECUTE format(
        'CREATE INDEX IF NOT EXISTS pipeline_build_events_%s_build_id 
            ON pipeline_build_events_%s (build_id)', NEW.id, NEW.id);

  EXECUTE format(
         'CREATE UNIQUE INDEX IF NOT EXISTS pipeline_build_events_%s_build_id_event_id 
           ON pipeline_build_events_%s (build_id, event_id)', NEW.id, NEW.id);

   RETURN NULL;
```

Reply:
Yep! Again, didn't really explain it so well, but that's the intention of having the `DeletePipeline` and `DeleteTeam` methods (rather than just having a general `Delete` method that acts on builds)

So, `DeletePipeline` would call `DROP pipeline_build_events_{id}`, and `DeleteTeam` would call `DROP team_build_events_{id}` and `DROP pipeline_build_events{id}` for each pipeline in the team